### PR TITLE
KeyBuilder: support function hashing

### PR DESCRIPTION
**Pull request by matthiasdiener**

### CODE REVIEW

File: pytools/test/test_persistent_dict.py [hunk 1]
```diff
@@ -719,6 +719,44 @@ def test_xdg_cache_home() -> None:
...
+    keyb = KeyBuilder()
+
+    assert keyb(global_fun) == keyb(global_fun) == \
+        "51b5980dd3a8aa13f6e83869e4a04c22973d7aaf96cb22899abdfdc55e15c9b2"
+    assert keyb(global_fun) != keyb(global_fun2)
+
+    def local_fun():
+        pass
+
+    def local_fun2():
```
Comment (inducer) [inline pytools/test/test_persistent_dict.py:740]: I wonder whether we should offer to hash local functions, given that they can be closures. For example:

```py
def get_fun(x):
    def add_x(y):
        return x + y
    return add_x
```

Would `gen_fun(5)` hash equal to `get_fun(7)`? Should it (probably not)?

Comment (matthiasdiener) [reply pytools/test/test_persistent_dict.py:740]: Yeah, they do hash equal currently (but the functions themselves do not compare equal).

Comment (matthiasdiener) [reply pytools/test/test_persistent_dict.py:773]: 6c8a26b improves this a bit, but there is still a small inconsistency (see 6aef85e).

Comment (inducer) [reply pytools/test/test_persistent_dict.py:773]: That's OK IMO. Hashing equal for not-equal objects is OK.

**Pull request merged**